quinn-udp/benches/throughput.rs

```diff
@@ -1,79 +1,120 @@
+use std::{
+    cmp::min,
+    io::{ErrorKind, IoSliceMut},
+    net::{Ipv4Addr, Ipv6Addr, UdpSocket},
+};
+
 use criterion::{criterion_group, criterion_main, Criterion};
+use tokio::{io::Interest, runtime::Runtime};
+
 use quinn_udp::{RecvMeta, Transmit, UdpSocketState};
-use std::cmp::min;
-use std::net::{Ipv4Addr, Ipv6Addr};
-use std::{io::IoSliceMut, net::UdpSocket, slice};
 
 pub fn criterion_benchmark(c: &mut Criterion) {
     const TOTAL_BYTES: usize = 10 * 1024 * 1024;
-    // Maximum GSO buffer size is 64k.
-    const MAX_BUFFER_SIZE: usize = u16::MAX as usize;
     const SEGMENT_SIZE: usize = 1280;
 
-    let (send_socket, send_state) = new_socket();
-    let (recv_socket, recv_state) = new_socket();
-    // Reverse non-blocking flag set by `UdpSocketState` to make the test non-racy
-    recv_socket.set_nonblocking(false).unwrap();
-
-    let max_segments = min(
-        UdpSocketState::new((&send_socket).into())
-            .unwrap()
-            .max_gso_segments(),
-        MAX_BUFFER_SIZE / SEGMENT_SIZE,
-    );
+    let mut rt = Runtime::new().unwrap();
+    let (send_socket, send_state) = new_socket(&mut rt);
+    let (recv_socket, recv_state) = new_socket(&mut rt);
     let dst_addr = recv_socket.local_addr().unwrap();
 
-    let mut receive_buffer = vec![0; MAX_BUFFER_SIZE];
-    let mut meta = RecvMeta::default();
+    let mut permutations = vec![];
+    for gso_enabled in [
+        false,
+        #[cfg(any(target_os = "linux", target_os = "windows"))]
+        true,
+    ] {
+        for gro_enabled in [false, true] {
+            #[cfg(target_os = "windows")]
+            if gso_enabled && !gro_enabled {
+                // Windows requires receive buffer to fit entire datagram on GRO
+                // enabled socket.
+                //
+                // OS error: "A message sent on a datagram socket was larger
+                // than the internal message buffer or some other network limit,
+                // or the buffer used to receive a datagram into was smaller
+                // than the datagram itself."
+                continue;
+            }
 
-    for gso_enabled in [false, true] {
-        let mut group = c.benchmark_group(format!("gso_{}", gso_enabled));
-        group.throughput(criterion::Throughput::Bytes(TOTAL_BYTES as u64));
+            permutations.push((gso_enabled, gro_enabled));
+        }
+    }
 
-        let segments = if gso_enabled { max_segments } else { 1 };
-        let msg = vec![0xAB; SEGMENT_SIZE * segments];
+    for (gso_enabled, gro_enabled) in permutations {
+        let mut group = c.benchmark_group(format!("gso_{}_gro_{}", gso_enabled, gro_enabled));
+        group.throughput(criterion::Throughput::Bytes(TOTAL_BYTES as u64));
 
+        let gso_segments = if gso_enabled {
+            send_state.max_gso_segments()
+        } else {
+            1
+        };
+        let msg = vec![0xAB; min(MAX_DATAGRAM_SIZE, SEGMENT_SIZE * gso_segments)];
```

Comment:
Why was this `min` introduced? It doesn't seem related to the intent of the commit.

Reply:
The `min` is need on Linux, which requires a segmented UDP datagram not to exceed the UDP datagram size limit, i.e. ~64k minus headers.

You are right, this does not directly relates to the intent of https://github.com/quinn-rs/quinn/pull/2010/commits/160a22ca0c071308c677d01aaa53c6ce9f34732f. I extracted it into a separate commit. 

See https://github.com/quinn-rs/quinn/pull/2010/commits/51ef00cc8b36103b1c88174042d67d355f266f96 and commit description for details.